components/SurveyModal.js

```diff
@@ -0,0 +1,566 @@
+          name="FNAME"
+          onChange={event => updateForm(event, 'firstName')}
+        />
+        <input
+          style={{ marginBottom: '0.5em' }}
+          type="text"
+          placeholder="Last Name"
+          name="LNAME"
+          onChange={event => updateForm(event, 'lastName')}
+        />
+        <input
+          style={{ marginBottom: '0.5em' }}
+          type="email"
+          placeholder="Email"
+          name="EMAIL"
+          onChange={event => updateForm(event, 'email')}
+        />
+      </fieldset>
+      <div className="tji-modal__actions">
+        <button type="button" onClick={() => skipStep()} className="btn--simple">
+          No Thanks
+        </button>
+        <button
+          type="button"
+          className="btn btn--primary"
+          onClick={() => validateStep(requiredFields, newsletterCallback)}
+        >
+          Continue
+        </button>
+      </div>
+      {stepError !== '' && <p className="tji-modal__form__error">{stepError}</p>}
+    </form>
+  );
+}
+
+Step4.propTypes = {
+  skipStep: PropTypes.func.isRequired,
+  validateStep: PropTypes.func.isRequired,
+  updateForm: PropTypes.func.isRequired,
+  stepError: PropTypes.string,
+};
+
+function Step5(props) {
+  const { cancelForm } = props;
+  return (
+    <form className="tji-modal__form">
+      <h2 className="tji-modal__title">Before you go...</h2>
+      <p className="tji-modal__description">
+        TJI provides its data to all users for free. Please consider making a contribution so we can continue to be a
+        free resource.
+      </p>
+      <div className="tji-modal__actions">
+        <button type="button" onClick={() => cancelForm()} className="btn--simple">
+          Not Now
+        </button>
+        <button type="button" className="btn btn--primary" onClick={() => (window.location.href = 'donate')}>
+          Donate!
+        </button>
+      </div>
+    </form>
+  );
+}
+
+Step5.propTypes = {
+  cancelForm: PropTypes.func.isRequired,
+};
+
+class SurveyModal extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      formActive: null,
+      currentStep: 1,
+      stepError: '',
+      surveyData: {
+        whoami: '',
+        dataSought: '',
+        dataFound: '',
+        firstName: '',
+        lastName: '',
+        email: '',
+        amount: 5,
+      },
+    };
+
+    this.validateStep = this.validateStep.bind(this);
+    this.cancelForm = this.cancelForm.bind(this);
+    this.updateForm = this.updateForm.bind(this);
+    this.skipStep = this.skipStep.bind(this);
+  }
+
+  componentDidMount() {
+    const dataDownloaded = localStorage.getItem('dataDownloaded') === 'true';
+
+    this.setState({
+      formActive: !dataDownloaded,
+    });
+
+    localStorage.setItem('dataDownloaded', 'true');
+  }
+
+  // Validate required fields and either move to the next step or add an error message
+  validateStep(requiredFields = [], callback = () => {}) {
```

Comment:
I like this approach for validation!

One thing I've appreciated in form validation is when the form makes it clear specifically which fields are invalid. What do you think about incorporating that somehow? Do you think we could get any of that for cheap by leveraging HTML5 validation and the `required` attribute on the `<input>` elements?

Reply:
I updated this to incorporate HTML5 validation as well. I agree, it's great to have both methods involved.

Reply:
Nice! Looks great.